### PR TITLE
Compaction continues as possible and warns if it might not delete enough entries

### PR DESCRIPTION
**Pull request by xirc**
This PR addresses the following (if compaction might not delete enough entries):
1. compaction will continue
2. compaction warns as possible

Compaction calculates estimated compacted log size using (the old `eventSourcingIndex`, which is fetched on the previous `SnapshotTick`). In this nature, if the log continues to grow much faster, the estimated compacted log size might be always greater than or equal to `compaction.log-size-threshold`. In this case, compaction won't happen and the log continues to grow, which consumes much memory. However, if compaction continues in such cases, it sends many snapshots to the snapshot-store every `SnapshotTick`, which might overwhelm the snapshot-store. It might be a trade-off at the time of writing. We might be able to choose either one, this PR chooses the reducing memory (we can choose the other one).

Either way, it is great to warn such cases. This PR attempt to warn as possible (not complete). This warning might be false-positive on the first `SnapshotTick`, which might be acceptable. If the warning continues, we should consult settings related to compaction and event sourcing.

### CODE REVIEW

File: src/main/scala/lerna/akka/entityreplication/raft/RaftActor.scala [hunk 1]
```diff
@@ -465,31 +465,34 @@ private[raft] class RaftActor(
...
           )
         }
-      } else if (snapshotSynchronizationIsInProgress) {
+      }
+      if (snapshotSynchronizationIsInProgress) {
         // Snapshot updates during synchronizing snapshot will break consistency
         if (log.isInfoEnabled)
           log.info("Skipping compaction because snapshot synchronization is in progress")
-      } else {
+      } else if (currentData.hasAppliedLogEntries) {
```
Comment (negokaz) [inline src/main/scala/lerna/akka/entityreplication/raft/RaftActor.scala:495]: [IMO] 
@xirc 
I think there is no need to log "Skipping compaction..." if it is obvious that the compaction will be skipped.
Would you like to change the code structure as follows?

```scala
if (currentData.hasAppliedLogEntries) {
  if (snapshotSynchronizationIsInProgress) {
    if (log.isInfoEnabled)
      log.info("Skipping compaction ...")
  } else {
    // start compaction...
  }
}
```

Comment (negokaz) [reply src/main/scala/lerna/akka/entityreplication/raft/RaftActor.scala:495]: I would like to withdraw this for more consideration.

Comment (negokaz) [reply src/main/scala/lerna/akka/entityreplication/raft/RaftActor.scala:495]: I would like to withdraw this for more consideration.

File: src/test/scala/lerna/akka/entityreplication/raft/RaftActorSpec.scala [hunk 2]
```diff
@@ -105,10 +105,6 @@ class RaftActorSpec extends TestKit(ActorSystem()) with RaftActorSpecBase {
         leaderCommit = applicableIndex,
       )
 
-      // To ensure compaction starts, CommitLogStore should handle AppendCommittedEntries.
-      commitLogStore.expectMsg(CommitLogStoreActor.AppendCommittedEntries(shardId, Seq.empty))
-      commitLogStore.reply(CommitLogStoreActor.AppendCommittedEntriesResponse(applicableIndex))
-
```
Comment (negokaz) [inline src/test/scala/lerna/akka/entityreplication/raft/RaftActorSpec.scala:111]: [nits] Can we remove also the declaration of `val commitLogStore`?

https://github.com/lerna-stack/akka-entity-replication/blob/ea8e4ed503020f018ff1626f361eb5d2c5ce89b3/src/test/scala/lerna/akka/entityreplication/raft/RaftActorSpec.scala#L78

Comment (xirc) [reply src/test/scala/lerna/akka/entityreplication/raft/RaftActorSpec.scala:111]: @negokaz Thanks! I've removed such code (e27f5b9).

**Review by negokaz — APPROVED** [review]
Great! LGTM 👍

**Pull request merged**